assertk/src/jvmMain/kotlin/assertk/assertions/any.kt

```diff
@@ -81,7 +81,7 @@ fun <T : Any> Assert<T>.isDataClassEqualTo(expected: T) = given { actual ->
 
 private fun <T> Assert<T>.isDataClassEqualToImpl(expected: T, kclass: KClass<*>?): Unit = given { actual ->
     if (actual == expected) return
-    val compareProps = actual != null && expected != null
+    val compareProps = actual != null && expected?.let { it::class } == kclass
     if (compareProps && kclass != null && kclass.isData) {
```

Comment:
This technically removes the `expected != null` check, but it's essentially still done since we check for equality with `kclass` which is then later checked not to be `null`.